shell/server/hack-session.js

```diff
@@ -323,10 +323,26 @@ HackSessionContextImpl.prototype.generateApiToken = function (petname, userInfo,
 };
 
 Meteor.methods({
-  newApiToken: function (grainId, petname, roleAssignment, forSharing, destroyIfNotUsedByTime) {
-    // Create a new user-oriented API token.
-    if (!this.userId) {
-      throw new Meteor.Error(403, "Must be logged in to create an API token.");
+  newApiToken: function (grainId, petname, roleAssignment, forSharing, expiresIfUnusedDuration,
+                         rawParentToken) {
+    // Creates a new UiView API token. If `rawParentToken` is set, creates a child token.
+    var userId = this.userId;
+    var parentToken;
+    if (!userId) {
+      if (!rawParentToken) {
+        throw new Meteor.Error(403, "If you are not logged in, you must specify a parent token to " +
+                                    "create a new API token.");
+      } else {
+        check(rawParentToken, String);
+        parentToken = Crypto.createHash("sha256").update(rawParentToken).digest("base64");
+        var parentApiToken = ApiTokens.findOne({_id: parentToken, grainId: grainId,
+                                                objectId: {$exists: false}});
+        if (!parentApiToken) {
+          throw new Meteor.Error(403, "No such parent token found.");
+        }
+        userId = parentApiToken.userId;
+        forSharing = parentApiToken.forSharing;
```

Comment:
Arguably it makes sense for a child to set `forSharing = true` when the parent had it `false`. This would happen e.g. when Alice creates an API token for use by an app acting on her behalf (forSharing = false), and then the app shares it to another user (forSharing = true).

OTOH, a child token should NOT be able to set forSharing = false when the parent has it true. (Or rather, if it does, this should be ignored.)

Abstractly, the `forSharing` flag on any link disassociates the consumer from any upstream identity.